File: original_php_project/src/User/Repository/UserPromocodeRepository.php
<?php

declare(strict_types=1);

namespace App\User\Repository;

use App\User\Entity\Promocode;
use App\User\Entity\UserPromocode;
use Doctrine\ORM\EntityManagerInterface;

final class UserPromocodeRepository
{

    public function __construct(
        private EntityManagerInterface $em
    ) {
    }

    public function countNumberOfUses(string $email, int $promocodeId): int
    {
        $qb = $this->em->getConnection()->createQueryBuilder();
        $qb->select('COUNT(1)');
        $qb->from('user_promocode');

        $qb->andWhere('email = :email');
        $qb->setParameter('email', $email);

        $qb->andWhere('promocode_id = :promocodeId');
        $qb->setParameter('promocodeId', $promocodeId);

        $qb->andWhere('state = :state');
        $qb->setParameter('state', UserPromocode::STATE_USED);

        return (int) $qb->execute()->fetchOne();
    }
}

File: target_java_project/src/main/java/com/goodhelp/user/domain/repository/UserPromocodeRepository.java
package com.goodhelp.user.domain.repository;

import com.goodhelp.user.domain.model.UserPromocode;
import com.goodhelp.user.domain.model.UserPromocodeState;

import java.util.Optional;

/**
 * Repository for user promocode usage.
 */
public interface UserPromocodeRepository {

    long countByEmailAndPromocodeIdAndState(String email, Long promocodeId, UserPromocodeState state);

    UserPromocode save(UserPromocode userPromocode);

    Optional<UserPromocode> findById(Long id);
}

